Reimplement gridicons using SVG external content.

sgomes opened the pull request:
Reimplement gridicons using an SVG external content approach, instead of SVG transpiled into React components. This helps keep images as images (which is good for performance), while maintaining styleability. Old browsers are supported through the already-included svg4everyone polyfill.

This PR supersedes and replaces #32172.

See #32364 for a similar change, which added Material Icons with the same approach.

#### Changes proposed in this Pull Request

* Remove AsyncGridicons component.
* Add Gridicons component, using offset data and SVG sprite from `gridicons` NPM package.
* Bump gridicons NPM package version to one that includes the new files.
* Modify gridicon CSS rules to be compatible with this approach.

#### Testing instructions

* Open the calypso.live branch of this PR in one browser tab.
* Open production in another browser tab.
* Alternate between both tabs, comparing as many different Calypso pages as you see fit. They should all be identical, pixel for pixel.

sgomes committed:
Reimplement gridicons using SVG external content.

sgomes committed:
Replace external Gridicon example w/ internal one.

sgomes committed:
Update `icons.md` with new gridicons info.

sgomes committed:
Replace usage of 'gridicons/dist/...'

sgomes commented:
Blocking this PR on #32750, as there's a bit of duplicate work between the two.

sgomes committed:
Enable SVG file-loader transform in gridicons.

jsnajdr commented on client/components/tinymce/plugins/media/advanced/index.jsx:
```diff
@@ -70,8 +70,9 @@ export default function( editor ) {
 			this.innerHtml(
 				ReactDomServer.renderToStaticMarkup(
+					// eslint-disable-next-line jsx-a11y/no-interactive-element-to-noninteractive-role
 					<button type="button" role="presentation" tabIndex="-1">
```
Removing the `role="presentation"` attribute looks like a better way to fix the a11y warning. I'm wondering why @aduth added it (long time ago in #2085) in the first place.

sgomes replied:
Fixed!

aduth replied:
> Removing the `role="presentation"` attribute looks like a better way to fix the a11y warning. I'm wondering why @aduth added it (long time ago in #2085) in the first place.

If I had to guess, it was likely to mimic markup otherwise rendered by a TinyMCE button in this context.

In retrospect, I ought to have made this more clear.

jsnajdr commented on assets/stylesheets/_vendor.scss:
```diff
@@ -5,15 +5,15 @@
 	fill: currentColor;
 
-	&.needs-offset g {
+	&.needs-offset use {
```
If the `svg4everybody` polyfill is active, there won't be any `use` elements to style. The polyfill will remove them and replace with the SVG markup it downloads with XHR.

We need a style that targets both `g` and `use` elements.

sgomes replied:
Excellent catch, thanks @jsnajdr !

I have pushed a fix for this using `:first-child`. Waiting for `calypso.live` before testing on IE11.

sgomes replied:
The fix appears to be working, from what I can tell.

jsnajdr reviewed (commented):
There's a little glitch with `needs-offset` and IE11.

I also have one question: when using the `use xlink:href` reference, can we target elements inside the SVG with CSS? I think we might use it at a few places, I'm not sure.

sgomes commented:
> I also have one question: when using the `use xlink:href` reference, can we target elements inside the SVG with CSS? I think we might use it at a few places, I'm not sure.

We can't, no; it's effectively a shadow DOM.

I couldn't find any instances of this targeting other than for alignment, which I rewrote in CSS.

jsnajdr commented:
> I couldn't find any instances of this targeting other than for alignment, which I rewrote in CSS.

Yes, I also found only instances where we style the top-level `g` element, usually changing color or offset. These selectors seems to be changed to target `use` everywhere.

sgomes committed:
Change icon adjustment CSS to use :first-child.

This ensures that alignment works both in polyfilled and
non-polyfilled setups.

sgomes committed:
Remove role=presentation from button.

jsnajdr commented on assets/stylesheets/_vendor.scss:
```diff
@@ -5,15 +5,15 @@
 	fill: currentColor;
 
-	&.needs-offset g {
+	&.needs-offset :first-child {
```
When the whole SVG is inline rather than referenced, doesn't this also offset all nested elements that happen to be the first child of their parent?
```
&.needs-offset > g:first-child, &.needs-offset > use:first-child {
  transform: ...
}
```
should be more reliable. What do you think?

sgomes replied:
Good catch, once again! But if the concern is to restrict to immediate descendants (which is a valid concern), then the following should suffice:

```css
&.needs-offset > :first-child {
  transform: ...
}
```

Would you prefer for me to still explicitly mention `g` and `use`?

jsnajdr replied:
> Would you prefer for me to still explicitly mention g and use?

Not really needed, just more safe and explicit IMO 🙂

sgomes replied:
No worries, I'l switch to that, then :)

sgomes committed:
Fix scoping bug in updated CSS rules.

jsnajdr reviewed (approved):
:+1: :tada:

sgomes committed:
Make CSS rules more explicit
